packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -0,0 +1,247 @@
+
+      files.push({
+        executeBackup: async handler => {
+          await Promise.all([
+            handler.outputFile(fileName, JSON.stringify(data, null, 2)),
+            handler.outputFile(
+              metaDataFileName,
+              JSON.stringify(
+                {
+                  jobId: job.id,
+                  scheduleId: schedule.id,
+                  timestamp,
+                },
+                null,
+                2
+              )
+            ),
+          ])
+          $defer.onFailure(async () => {
+            await Promise.all([
+              handler.unlink(fileName),
+              handler.unlink(metaDataFileName),
+            ])
+          })
+        },
+        dir,
+        retention: retentionXoMetadata - 1,
+      })
+    }
+    if (!isEmptyPools && retentionPoolMetadata > 0) {
+      const retention = retentionPoolMetadata - 1
+      files.push(
+        ...(await Promise.all(
+          poolIds.map(async id => {
+            const dir = `${scheduleDir}/${id}`
+            const fileName = `${dir}/${timestamp}.xml`
+            const metaDataFileName = `${dir}/${timestamp}.metadata.json`
+
+            const stream = await app.getXapi(id).getPoolMetadata(cancelToken)
+
+            return {
+              executeBackup: async handler => {
+                const [outputStream] = await Promise.all([
+                  handler.createOutputStream(fileName),
+                  handler.outputFile(
+                    metaDataFileName,
+                    JSON.stringify(
+                      {
+                        jobId: job.id,
+                        scheduleId: schedule.id,
+                        poolId: id,
+                        timestamp,
+                      },
+                      null,
+                      2
+                    )
+                  ),
+                ])
+                stream.pipe(outputStream)
+                $defer.onFailure(async () => {
+                  await Promise.all([
+                    handler.unlink(fileName),
+                    handler.unlink(metaDataFileName),
+                  ])
+                })
+              },
+              dir,
+              retention,
+            }
+          })
+        ))
+      )
+    }
+
+    if (files.length === 0) {
+      throw new Error('no retentions corresponding to the metadata modes found')
+    }
+
+    cancelToken.throwIfRequested()
+
+    return asyncMap(remoteIds, async id => {
+      const handler = await app.getRemoteHandler(id)
+      return files.map(async ({ executeBackup, dir, retention }) => {
+        // deleting old backups
+        await handler.list(dir).then(
+          list => {
+            if (retention > 0) {
+              list = list.slice(0, -retention)
+            }
+            return Promise.all(
+              list.map(fileToDelete => handler.unlink(`${dir}/${fileToDelete}`))
+            )
+          },
+          () => {}
```

Comment:
Don't do this, use `promise-toolbox/ignoreErrors` if you need to, otherwise you may hide some coding errors.